DR-2993: Utilize new TDR API Column Stats endpoint for Dataset filtering on snapshot create

snf2ye opened the pull request:
Changes deployed here: https://jade-sh.datarepo-dev.broadinstitute.org/ 
Please try to break 😄 
Ticket: https://broadworkbench.atlassian.net/browse/DR-2993
_____________________________________________________

https://github.com/DataBiosphere/jade-data-repo-ui/assets/13254229/23604c6c-233d-4145-b513-58ec180106eb


### The Changes
* Utilize Column stats TDR endpoint rather than directly querying BQ
* If we don filter on a field due to data type or array status, a message has been added to indicate why the user cannot filter on the filed. 
<img width="580" alt="image" src="https://github.com/DataBiosphere/jade-data-repo-ui/assets/13254229/87b98705-e8f6-42e0-92c2-26b285f762c7">

* We now display both the filtered and total count for checkbox text fields
<img width="581" alt="image" src="https://github.com/DataBiosphere/jade-data-repo-ui/assets/13254229/f6756058-86b6-43e7-8706-238577575ea1">


**There are 3 types of fields we can filter on:**
1. Text checkbox
 * Handles options with 30 or fewer unique text options
 * We track the total count and the filtered count of each of these unique fields
 * On applying a filter, we re-fetch the column stats and update the counts of each value. We continue to display all options if it has been filtered out of the data, we just set the filtered count to 0.
2. Free text
* We fetch all unique values upon opening the filter window. We do not update these on applying the filter. (This is a change -[ see comment](https://github.com/DataBiosphere/jade-data-repo-ui/pull/1417#discussion_r1195043891)) 
3. Numeric
* We fetch the min and max numeric value for the field upon opening the filter window. We do not update this on applying the filter. 

### Approach
I've tried to keep the changes as minimal as possible to help with the review process.  

Potential follow-on changes:
* I would like to add a follow-on PR that updates these components to typescript and into functional components. I have these changes in a separate branch. 
* Any remaining BQ references - Like removing scope? Anything else?
* Handle array fields (https://broadworkbench.atlassian.net/browse/DR-1476)
* Handle additional data types - date/time fields, boolean

### Helpful resources
[Updated understanding of useEffect](https://react.dev/learn/you-might-not-need-an-effect)

snf2ye committed:
Typescriptify + Functional Component: RangeInput

snf2ye committed:
Use Column stats endpoint for both free text and multi-select text fields

snf2ye committed:
wire up numeric handling

snf2ye committed:
working! 🎉

snf2ye committed:
Don't run on arrays for now

snf2ye committed:
lint/Cleanup/remove unused variables

lint

clean out unused variables

clean up

cleanup

- lint
- remove "originalValues"

snf2ye committed:
Remove bigquery code; Add filtering capability with col stats enpdoint

fix

add back join statement

wip

snf2ye committed:
Add filtering

lint

snf2ye committed:
fix endless loop

snf2ye commented:
Query builder tests are failing because we have not yet merged the API change to add the column stats endpoint. Manual testing of these functionalities works.

snf2ye committed:
rename tdrApiFilterStatement to filterStatement

snf2ye commented on src/components/common/data/DataView.tsx:
```diff
@@ -99,7 +99,6 @@ function DataView({
   // Can be removed after DR-2483
   const showPanels = panels.length > 0;
-  // Only used for Direct BQ Query
   const isDatasetFiltered = filterStatement.length > 0;
```
I originally added this comment, but I'm not sure why - It's not accurate as far as I can tell.

snf2ye commented on src/components/dataset/data/sidebar/filter/FreetextFilter.jsx:
```diff
@@ -35,54 +33,20 @@ export class FreetextFilter extends React.PureComponent {
     this.setState({
       inputValue: value,
     });
-
-    const response = await bq.getAutocompleteForColumnDebounced(
-      value,
-      column.name,
-      dataset,
-      tableName,
-      token,
-      filterStatement,
-      joinStatement,
-    );
-
-    const transformedResponse = this.transformResponse(response);
-    this.setState({
-      options: transformedResponse,
-    });
```
I don't think this is necessary. 
By filtering down the responses, we remove the ability to add more options after filtering. 
Let's say there are 3 options: 'a' 'b', & 'c'. (This is a bad example because it would actually use the checkbox option, which DOES keep the original values in the list even after filtering with an updated count according to the filter set. Anyway...)
* let's say the user selects 'a' and then hits "apply filter"
* Then, the user decides they also want to include 'b'. If we kept this code in place, the the option 'b' would no longer be present. Yes, they could just type in 'b' if they knew what they were looking for, but otherwise, they would have to clear the filter and then go back and select 'a' and 'b'.

nmalfroy replied:
I agree.  this is a good change.  I do worry about not debouncing though.  I want to make sure we're not hammering BQ

nmalfroy replied:
But I'll add that click testing, it seems to not be overly aggressive so I think it's fine

snf2ye replied:
It should only ever hit BQ (via API) on first load and on actually hitting the "apply filter" button. With this, do you think we're at risk for hammering the BQ?

snf2ye commented on src/components/dataset/data/sidebar/DataViewSidebarItem.jsx:
```diff
@@ -25,13 +25,9 @@ export class DataViewSidebarItem extends React.PureComponent {
   static propTypes = {
     classes: PropTypes.object,
     column: PropTypes.object,
-    dataset: PropTypes.object,
     filterData: PropTypes.object,
-    filterStatement: PropTypes.string,
     handleChange: PropTypes.func,
-    joinStatement: PropTypes.string,
     tableName: PropTypes.string,
-    token: PropTypes.string,
```
We no longer need to pass these items through:
- Dataset & filterStatement have been moved to redux state
- joinStatement is not needed for the column stats endpoint
- token was only used when directly talking to bq

snf2ye commented on src/components/dataset/data/sidebar/filter/CategoryFilter.jsx:
```diff
@@ -36,7 +39,7 @@ export class CategoryFilter extends React.PureComponent {
           />
         }
-        label={`${name} (${count})`}
+        label={`${name} (${formattedCount})`}
```
Slight update to the original functionality - this now displays both the filtered and total count. It originally just showed the filtered count. 
With change:
<img width="566" alt="image" src="https://github.com/DataBiosphere/jade-data-repo-ui/assets/13254229/6a9cb090-590b-4e58-9d19-0c34628ad452">

snf2ye commented on src/components/dataset/data/sidebar/filter/CategoryFilterGroup.jsx:
```diff
@@ -45,17 +45,24 @@ export class CategoryFilterGroup extends React.PureComponent {
-    const { column, values, table, originalValues, filterMap } = this.props;
-    const checkboxes = _.keys(originalValues).map((name) => {
-      const count = parseInt(_.get(values, name, '0'), 10);
+    const { column, values, table, filterMap, originalValues } = this.props;
+    const checkboxes = originalValues?.map((val) => {
+      const totalCount = val.count;
+      let filteredCount = 0;
+      values.map((filteredValue) => {
+        if (filteredValue.value === val.value) {
+          filteredCount = filteredValue.count;
+        }
+      });
```
(Explanation because this wasn't entirely clear to me before getting into this code)
For checkbox text fields (so, text fields with fewer than 30 options), we track both the full set of options ('originalValues') and the filtered set of options ('values'). The only thing this buys us is that we can display the count of the number of times we see a given value and update this count after filtering. 
We display all of the options (all of the 'originalValues'), even if filtering has eliminated an option from the data view. If an option has been eliminated, we display the count as 0. We get the filtered count from the list of filtered 'values'.

snf2ye commented on src/components/dataset/data/sidebar/filter/CategoryWrapper.tsx:
```diff
@@ -0,0 +1,89 @@
+  filterStatement,
+  toggleExclude,
+}: CategoryWrapperProps) {
+  const refreshColumnStats =
+    column.values === undefined ||
+    (column?.originalValues && column.originalValues.length <= CHECKBOX_THRESHOLD);
+  useEffect(() => {
+    // only update for checkbox text columns on filtering
+    if (refreshColumnStats) {
+      dispatch(
+        getColumnStats(
+          ResourceType.DATASET,
+          dataset.id,
+          tableName,
+          column.name,
+          ColumnDataTypeCategory.TEXT,
+        ),
+      );
+    }
+  }, [dispatch, dataset.id, tableName, column.name, filterStatement, refreshColumnStats]);
```
Note: We originally talked about only fetching the column stats after a user clicks on a particular field in the filter panel in order to avoid extra trips to BQ/Synapse.  I decided to leave this functionality as is for now since it would be a change in the user experience(& a little complicated to get exactly right - for example, what happens when you filter? Do you then update all fields? Or again only update on field click?), but am glad to revisit this change. 

@nmalfroy  - I'd be curious for your thoughts on this after clicking around in the updated UI.

nmalfroy replied:
I think it works in the UI.  My main concern is for azure datasets (longer query times) with a lot of columns.  I think it's fair to keep the current behavior and see how it performs once you roll out the Azure backend

snf2ye replied:
Ah, you called it. With Azure, the get stats calls are pretty slow and when you have a bunch of columns, the filter becomes unusable. In the example dataset from our integration tests that has 50+ columns, it takes 2 **minutes** to retrieve all of the stats. So, I think we need to make this change before we enable it for Azure.

snf2ye replied:
And, actually, it's not just Azure. It totally stalls the UI for the "transcript_consequence" table in GCP. https://jade-sh.datarepo-dev.broadinstitute.org/datasets/5ed80a11-9c2e-402c-a583-fe4c85ead4b4

nmalfroy replied:
Ah yeah, I see what you mean...So basically, we track expanded columns and any time a filter changes, only call the requests for the expanded columns?

snf2ye committed:
Update and add comments

snf2ye committed:
lint

snf2ye committed:
handle empty table

snf2ye committed:
display message for empty numeric column

snf2ye committed:
Forces columns to update when the table switches back to first table on page timeout

snf2ye committed:
lint

samanehsan reviewed (approved):
This looks great! I did some manual testing on your deployment and the filtering worked as expected 😄 It would be great to add tests for the individual components as well, like the `RangeFilter`

snf2ye committed:
Fix test

nmalfroy commented on src/reducers/query.ts:
```diff
@@ -146,6 +154,49 @@ export default {
+          polling: { $set: false },
+        }),
+      [ActionTypes.COLUMN_STATS_TEXT_SUCCESS]: (state, action: any) => {
+        const values = action.payload.queryResults.data.values;
```
Might make sense to just do this now so it doesn't keep squawking at us
```suggestion
        const { values } = action.payload.queryResults.data;
```

snf2ye replied:
Ah, thanks!

nmalfroy reviewed (approved):
Looks great! And click testing works awesome (awesomely?)

nmalfroy reviewed (approved):
Looks great! And click testing works awesome (awesomely?)

nmalfroy reviewed (approved):
Looks great! And click testing works awesome (awesomely?)

Also thanks for the typescriptification!

snf2ye committed:
Update src/reducers/query.ts

Co-authored-by: Nick Malfroy-Camine <nmalfroy@broadinstitute.org>

snf2ye committed:
Only filter on field expansion

snf2ye committed:
lint

snf2ye committed:
flip expanded var

snf2ye committed:
Fix bug: FIltering disappeared when filtered down to 0 rows

snf2ye committed:
Fix the pause in loading between all  & filtered retrieval

wip

snf2ye committed:
Fixed the loading issues; null error in console;

snf2ye committed:
lint

snf2ye committed:
Needed to set filterHasUpdated var for all filter types

snf2ye committed:
Only load numeric stats on first expansion of field

snf2ye committed:
Fix lint

snf2ye committed:
doh, minValue

snf2ye committed:
Add CategoryWrapper test

snf2ye committed:
Add RangeFilter Test; lint fixes

snf2ye commented:
>  It would be great to add tests for the individual components as well, like the `RangeFilter`

@samanehsan  - Good point! I've added component tests for RangeFilter and CategoryWrapper with a particular focus on the logic I've added to these two components.

snf2ye committed:
use original values for freetext dropdown

nmalfroy commented on src/components/dataset/data/sidebar/filter/CategoryWrapper.tsx:
```diff
@@ -47,9 +45,41 @@ function CategoryWrapper({
 
-  const { originalValues, values } = column;
+  // Run when filter has been updated
+  useEffect(() => {
```
This is great.  It made a very noticeable difference in the UI

nmalfroy commented on src/components/dataset/data/sidebar/filter/CategoryFilter.jsx:
```diff
@@ -33,7 +33,7 @@ export class CategoryFilter extends React.PureComponent {
             checked={checked}
             onChange={this.handleChange}
-            value={name}
+            value={name ?? "null"}
```
Does it make sense to have this be the italic `(empty)` like we show in the table?

samanehsan replied:
I agree, I think "Empty" or "None" would be clearer!

snf2ye replied:
Great! I will cover this in my next UI PR!

nmalfroy reviewed (approved):
Awesome!  Just a couple of comments that could totally be followons

samanehsan commented on src/components/dataset/data/sidebar/filter/CategoryWrapper.test.tsx:
```diff
@@ -0,0 +1,189 @@
+      values: [{ value: 'a', count: 1 }],
+      originalValues: [
+        { value: 'a', count: 1 },
+        { value: 'b', count: 2 },
```
I'm not sure I understand, why does this test contain repeated values?

snf2ye replied:
We don't have anything on the UI side that checks if the values are unique (We expect the API to return a list of unique values). For this test, I'm trying to just get more than 30 elements in the values list so that it switches from the checkbox display to the freetext dropdown. But, it is unclear- I'll make them unique.

samanehsan replied:
Oh that makes sense now!

samanehsan reviewed (approved):
Thanks for adding in the tests! (I just had a question about one of them)

snf2ye committed:
update test values to be unique
